src/main/java/io/vertx/ext/mail/impl/dkim/DKIMSigner.java

```diff
@@ -0,0 +1,491 @@
+          final String tobeSigned = headersToSign(encodedMessage).append(dkimSignHeaderCanonic).toString();
+          if (logger.isDebugEnabled()) {
+            logger.debug("To be signed DKIM header: " + tobeSigned);
+          }
+          String returnStr;
+          synchronized (signatureService) {
+            signatureService.update(tobeSigned.getBytes());
+            String sig = Base64.getEncoder().encodeToString(signatureService.sign());
+            returnStr = dkimTagListBuilder.append(sig).toString();
+          }
+          if (logger.isDebugEnabled()) {
+            logger.debug(DKIM_SIGNATURE_HEADER + ": " + returnStr);
+          }
+          promise.complete(returnStr);
+        } catch (Exception e) {
+          promise.fail(e);
+        }
+      } else {
+        logger.warn("Failed to calculate the body hash", bhr.cause());
+        promise.fail(bhr.cause());
+      }
+    });
+    return promise.future();
+  }
+
+  // the attachPart is a base64 encoded stream already when this method is called.
+  private void walkThroughAttachStream(MessageDigest md, ReadStream<Buffer> stream, AtomicInteger written, Promise<Void> promise) {
+    final Pipe<Buffer> pipe = stream.pipe();
+    Promise<Void> pipePromise = Promise.promise();
+    pipePromise.future().setHandler(pr -> {
+      pipe.close();
+      if (pr.succeeded()) {
+        promise.complete();
+      } else {
+        promise.fail(pr.cause());
+      }
+    });
+    pipe.to(new WriteStream<Buffer>() {
+      private AtomicBoolean ended = new AtomicBoolean(false);
+
+      @Override
+      public WriteStream<Buffer> exceptionHandler(Handler<Throwable> handler) {
+        return this;
+      }
+
+      @Override
+      public Future<Void> write(Buffer data) {
+        Promise<Void> promise = Promise.promise();
+        write(data, promise);
+        return promise.future();
+      }
+
+      @Override
+      public void write(Buffer data, Handler<AsyncResult<Void>> handler) {
+        if (!ended.get() && !digest(md, data.getBytes(), written)) {
+          // can be end now
+          ended.set(true);
+        }
+        if (handler != null) {
+          handler.handle(Future.succeededFuture());
+        }
+      }
+
+      @Override
+      public void end(Handler<AsyncResult<Void>> handler) {
+        ended.compareAndSet(false, true);
+        if (handler != null) {
+          handler.handle(Future.succeededFuture());
+        }
+      }
+
+      @Override
+      public WriteStream<Buffer> setWriteQueueMaxSize(int maxSize) {
+        return this;
+      }
+
+      @Override
+      public boolean writeQueueFull() {
+        return false;
+      }
+
+      @Override
+      public WriteStream<Buffer> drainHandler(@Nullable Handler<Void> handler) {
+        return this;
+      }
+    }, pipePromise);
+  }
+
+  private boolean digest(MessageDigest md, byte[] bytes, AtomicInteger written) {
```

Comment:
multiple calls are made here, so if the `md` is accessed by multiple threads it will cause errors as `MessageDigest` is not thread safe. Perhaps synchronize the method or the `md` object itself.

Reply:
The `md` instance here is executed in one Thread. Because the instance is created during the method:  [bodyHashing](https://github.com/vert-x3/vertx-mail-client/blob/cd91c8f8709d5c9f9043cd30bda48906fa4d97cb/src/main/java/io/vertx/ext/mail/impl/dkim/DKIMSigner.java#L370) call, and it will go through the email parts one by one in the same thread as the connection's context.